### PR TITLE
GEODE-1758: Refactor, test and clean up TypeUtils

### PR DESCRIPTION
**Pull request by jujoramos**
**Fix for GEODE-1758**

The pull request contains two different commits on purpose:

- The first one includes all JUnit tests for the `TypeUtils` class, basically to make sure that the changes won't break existing behavior nor introduce new bugs.
- The second one is the actual refactoring of the `TypeUtils` class without major modifications to the original tests, basically to make sure everything still works as expected.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/ObjectTypeAssert.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.assertj.core.api.AbstractAssert;
+
+import org.apache.geode.cache.query.types.ObjectType;
+
+/**
+ * Custom Assertion to validate {@link ObjectType} instances.
+ */
+public class ObjectTypeAssert extends AbstractAssert<ObjectTypeAssert, ObjectType> {
+
+  /**
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/types/ObjectTypeAssert.java]: Please delete all incomplete javadocs

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,1197 @@
...
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.pdx.internal.PdxInstanceEnum;
+import org.apache.geode.pdx.internal.PdxString;
+import org.apache.geode.test.junit.categories.UnitTest;
+
+@Category(UnitTest.class)
+@PowerMockIgnore("*.UnitTest")
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(TypeUtils.class)
+public class TypeUtilsJUnitTest {
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: We're currently naming unit tests FooTest and integration tests FooIntegrationTest. No need for "JUnit" in the name of the class.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,1197 @@
...
+              .isEqualTo(Boolean.FALSE);
+          assertThat(TypeUtils.compare(originDate, originDate2, OQLLexerTokenTypes.TOK_GE))
+              .isEqualTo(Boolean.TRUE);
+          assertThat(TypeUtils.compare(originDate, originDate2, OQLLexerTokenTypes.TOK_NE))
+              .isEqualTo(Boolean.FALSE);
+          verify((Comparable) originDate, times(0)).compareTo(any());
+          verify((Comparable) originDate2, times(0)).compareTo(any());
+          verify(temporalComparator, times(6)).compare(any(), any());
+          reset(temporalComparator);
+        } catch (TypeMismatchException typeMismatchException) {
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: Remove any try-catch blocks and simply add "throws Exception" to the signature of the test methods.

Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: Hey @kirklund 

Thanks for reviewing this. By default Java doesn't support throwing checked exceptions from a lambda expression, that's why I'm catching the checked exceptions and re-throwing them as unchecked exceptions in the tests. There are certain ways to [handle checked exceptions in Java streams](https://www.oreilly.com/ideas/handling-checked-exceptions-in-java-streams), but they're not trivial, and not even worth the hassle for test classes. If an exception is thrown the test will fail anyway, and the root cause of the exception will be shown as for every other test.
Let me know if you still want this to be changed.
Cheers.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java (hunk 2 above)
Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: Hey @kirklund,

In a previous pull request I was specifically asked by a reviewer to change the name of one of my tests from "FooTest" to "FooJUnitTest", that's why I create the tests using that suffix now. Moreover, within the [Code Contributions](https://cwiki.apache.org/confluence/display/GEODE/Code+contributions), specifically in the [Writing tests](https://cwiki.apache.org/confluence/display/GEODE/Writing+tests) section, we have the following:

```
JUnit tests should

. Use Junit 4 Syntax
. End with JUnitTest
. Contain an Category annotation of either UnitTest or IntegrationTest. UnitTests as should complete in milliseconds and test a specific class.
```

Which approach/guideline is the correct one?, are all reviewers on the same page regarding this?.
In the end, should I change the class name from `TypeUtilsJUnitTest` to `TypeUtilsTest` for this pull request to be accepted?. If the answer is "yes", shouldn't we change the Wiki and make sure all reviewers are in sync?.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/ObjectTypeAssert.java (hunk 1 above)
Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/ObjectTypeAssert.java]: 👍

**Comment by jujoramos** [comment]
Hello @jhuynh1,

Thanks for pointing this out, I totally forgot about the new serialization stuff... the `TypeUtils` class, and its inner classes as well, are only used as utility classes so they shouldn't be persisted nor sent through the wire at any point, meaning that they shouldn't be serialized and, thus, it should be safe to add them to `geode-core/out/test/resources/org/apache/geode/codeAnalysis/excludedClasses.txt`. I'll submit the changes shortly.
Thanks.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java (hunk 2 above)
Comment (kirklund) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: Hi Juan, I've brought this up on the dev-list.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java (hunk 3 above)
Comment (kirklund) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: Oops, you're right! I was thinking these were dunit vm invocations which you can remove the catch-blocks from lambda calls.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java (hunk 2 above)
Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: Hey @kirklund ,
Great!. I'll wait for the final decision and change the test name if needed.
Cheers.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java (hunk 3 above)
Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/types/TypeUtilsJUnitTest.java]: 👍

**Pull request merged**